### DISCUSSION TITLE
Crazyflie SDIO, UWB Beacon and wifi.

### DISCUSSION THREAD

**Discussion by Aerialmagic** [bitcraze/discussions#696] [Ideas]
I hope this meets you well , I don't know if exactly this is the right place to ask about this, I have some questions I will like to ask or maybe they are in a kind of asking for support, we are designing an all in one drone taking our reference from the 18650 battery 3d printed crazyflie bolt drone, the questions goes.

1. The crazyflie uses the microsd card spi protocol, is it possible to use the faster SDIO or QSPI protocol for faster data logging and reading by the crazyflie, as we are currently using an SPI flash chip in a QSPI configuration to be able to use it with Ardupilot in future?
2. We have in production a UWB Beacon that uses the DWM3000 together with the raspberry pi rp2040, it will calculate the positions on board and send the position information out through the UART, will the crazyflie stack be able to receive those and use that instead of the spi UWB?
3. And the last question, is there a part of crazyflie stack that makes it possible to use wifi connection instead of the 2.4GHz radio protocol? As wifi is the only communication protocol currently for the drone.

**Comment by ataffanel** [comment]
Hi,

I can try to answer in order

1. If I understand well, you are making your own drone and plan to use Ardupilot so, anything should be possible as long as you design it... The STM32F405 has a proper SDIO peripheral but, as far as I can tell glancing at the datasheet, no QSPI. The SDIO pins where unfortunatly impossible to route to the Crazyflie 2 expansion port. But, again, if you are making your own design and are planning to use Ardupilot, you are not limited to the STM32F405 and the Crazyflie pinout so you can get SDIO and/or QuadSPI
2. It is possible to add new positioning system to the Crazyflie yes. Since we support so many already, there is support in the firmware to add yet another one. The details will depends a lot of what kind of data you are getting but essencially you will need to implement a new input to the Kalman Filter. The current external position implementation could be used as a reference if what you are getting is position.
3. The AI deck adds wifi connectivity to the Crazyflie so this is something that has been done. The current strategy though is to offload all wifi and network connection handling to the wifi chip (an ESP32 in that case) and to talk to the Crazyflie over UART using the `PRX` and `CRTP` protocol.

**Reply by Aerialmagic** [reply]
Thank you very much, I'd like to clear out some things, as it is, we want to be able to use the crazyflie stack and also able to use Ardupilot stack as we have different use cases for them, currently we are working with the crazyflie as the first stack as our hardware uses the AT32F435 from Artery, the crazyflie bootloader is easier to work with as the Ardupilot ChibiOS is rather not, the drones are 3" , so with the MCU we are using, we are not limited to IOs , it will just take alot of fiddling around to get it working, that's why we are here, maybe you could point out some key places to touch regarding
* Changing the SPI SDcard to QSPI flash
* Changing the eeprom to f-ram
* Using external beacon with UART(kalman filter is understood) 
 to make our work easier, 
thank you.

**Comment by ataffanel** [comment]
Those are very generic questions and will depend a lot of the exact platform you are using. I have never used the microcontroller you are using so I am not sure exactly how this things will be implemented.

The SD-card is mostly handled by FatFS so you will have to port it to use sdio instead of spi. The eeprom is handled by eeprom.c/h and as I already pointed out we have some extenal position input. The simplest (X,Y,Z) kind is in the crtp_position_service.

What you are trying to do is a pretty complex endeavor and you will need to fully understand the architecture of the Crazyflie firmware so I suggest you look at the whole architecture to understand what might cause problem porting it to a new MCU. You are very welcome if you have any more precise questions about architecture or implementations. I suggest you create one new thread per questions/topic though.
